controller/src/main/java/io/pravega/controller/store/stream/tables/TableHelper.java

```diff
@@ -435,6 +418,135 @@ public static int getSegmentCount(final byte[] historyTable) {
         return indexStream.toByteArray();
     }
 
+    /**
+     * Method to check if a scale operation is currently ongoing.
+     * @param historyTable history table
+     * @param segmentTable segment table
+     * @return true if a scale operation is ongoing, false otherwise
+     */
+    public static boolean isScaleOngoing(final byte[] historyTable, final byte[] segmentTable) {
+
+        HistoryRecord latestHistoryRecord = HistoryRecord.readLatestRecord(historyTable, false).get();
+        return latestHistoryRecord.isPartial() || !latestHistoryRecord.getSegments().contains(getLastSegmentNumber(segmentTable));
+    }
+
+    /**
+     * Method to check if no scale operation is currently ongoing and scale operation can be performed with given input.
+     * @param segmentsToSeal segments to seal
+     * @param historyTable history table
+     * @return true if a scale operation can be performed, false otherwise
+     */
+    public static boolean canScaleFor(final List<Integer> segmentsToSeal, final byte[] historyTable) {
+        HistoryRecord latestHistoryRecord = HistoryRecord.readLatestRecord(historyTable, false).get();
+        return latestHistoryRecord.getSegments().containsAll(segmentsToSeal);
+    }
+
+    /**
+     * Method that looks at the supplied input and compares it with partial state in metadata store to determine
+     * if the partial state corresponds to supplied input.
+     * @param segmentsToSeal segments to seal
+     * @param newRanges new ranges to create
+     * @param historyTable history table
+     * @param segmentTable segment table
+     * @return true if input matches partial state, false otherwise
+     */
+    public static boolean isRerunOf(final List<Integer> segmentsToSeal,
```

Comment:
I think this is the method used to determine if a given scale event is already processed, or was left half way,  or is yet to be processed.

As we had discussed previously, this method can be simplified by observing that scale events shall be processed sequentially for a stream by the scale event processor.

Following is the logic we had discussed.

Let `EventStatus` be the enum `{Done, InProgress, Pending}`. We could use the following method to do this classification.
```
EventStatus classifyEvent(ScaleEvent scaleEvent) {
  HistoryRecord latest = HistoryTable.getLatestRecord();
  if (latest.segments.containsAll(scaleEvent.segmentsToSeal)) {
    return Pending;
  } else {
    if (latest.isPartial && HistoryTable.getActiveRecord().segments.containsAll(scaleEvent.segmentsToSeal)) {
      return InProgress;
    } else {
      return Done;
    }
  } 
}
```